Document Field Groups REST API and FIELD_GROUP_LIST_ON_GENERATE event (#531)
mautic/developer-documentation-new: docs/plugin_extensions/contacts.rst

      - Optional
      - callback
      - Callback to custom parse a result. This is optional and mainly used to handle a column result when all results are already looped over for ``$serializedColumns`` and $dateTimeColumns.

> Customizing Custom Field Groups
> ******************************

Reviewer: ```suggestion
.. vale off

Customizing Custom Field Groups
******************************

.. vale on
```

Mautic organizes Contact and Company Custom Fields into groups, which appear as tabs on the Contact and Company view and edit pages. Beyond the built-in groups, admins can create their own Custom Field Groups under :guilabel:`Settings → Custom Field Groups`.

> As Mautic builds the group list for a view, it dispatches the ``LeadEvents::FIELD_GROUP_LIST_ON_GENERATE`` event. Listen to this event to add your own groups or rename an existing group. Mautic resolves each group's display name through this event, so a Plugin can label custom groups without registering static translation keys.

Author: The `FIELD_GROUP_LIST_ON_GENERATE` constant is defined in `app/bundles/LeadBundle/LeadEvents.php`, and `FieldGroupModel::getTranslatedGroups()` dispatches a `FieldGroupListEvent` on that event before returning the group map — so listeners can add or rename groups. The display-name-resolution-without-translation-keys point comes from the PR description and `FieldGroupModel::getGroups()` using the entity name as the fallback display value.

Source: https://github.com/mautic/mautic/pull/16265/files

    use Mautic\LeadBundle\LeadEvents;
    use Symfony\Component\EventDispatcher\EventSubscriberInterface;
    use Symfony\Contracts\Translation\TranslatorInterface;

    final class FieldGroupSubscriber implements EventSubscriberInterface
    {
        public function __construct(private TranslatorInterface $translator)
        {
        }

        public static function getSubscribedEvents(): array
        {
            return [
                LeadEvents::FIELD_GROUP_LIST_ON_GENERATE => ['onFieldGroupGenerate', 0],
            ];
        }

        public function onFieldGroupGenerate(FieldGroupListEvent $event): void
        {
            // Only act on Contact (lead) field groups.
            if ('lead' !== $event->getObject()) {
                return;
            }

            $groups = $event->getGroups();

            // Add a new group, keyed by its alias.
            $groups['loyalty'] = $this->translator->trans('mautic.hello.world.field_group.loyalty');

            $event->setGroups($groups);
        }
    }

The ``FieldGroupListEvent`` provides the following methods:

.. list-table::
   :header-rows: 1

   * - Method

>      - Description

Author: The `getGroups()`/`setGroups()`/`getObject()` methods and the `[alias => displayName]` array shape (with `object` being `'lead'` or `'company'`) are taken directly from `app/bundles/LeadBundle/Event/FieldGroupListEvent.php`.

Source: https://github.com/mautic/mautic/pull/16265/files

      - Optional
      - callback
      - Callback to custom parse a result. This is optional and mainly used to handle a column result when all results are already looped over for ``$serializedColumns`` and $dateTimeColumns.

.. vale off

Customizing Custom Field Groups
******************************

.. vale on

> Mautic organizes Contact and Company Custom Fields into groups, which appear as tabs on the Contact and Company view and edit pages. Beyond the built-in groups, admins can create their own Custom Field Groups under **Settings** > **Custom Field Groups**.

Reviewer: ```suggestion
.. vale off

Mautic organizes Contact and Company Custom Fields into groups, which appear as tabs on the Contact and Company view and edit pages. Beyond the built-in groups, admins can create their own Custom Field Groups under **Settings** > **Custom Field Groups**.

.. vale on
```

> Mautic organizes Contact and Company Custom Fields into groups, which appear as tabs on the Contact and Company view and edit pages. Beyond the built-in groups, admins can create their own Custom Field Groups under :guilabel:`Settings → Custom Field Groups`.

Reviewer: No need to use `guilabel` to keep consistent.

```suggestion
.. vale off

Mautic organizes Contact and Company Custom Fields into groups, which appear as tabs on the Contact and Company view and edit pages. Beyond the built-in groups, admins can create their own Custom Field Groups under **Settings** > **Custom Field Groups**.

.. vale on
```
